### PR TITLE
chore: update beachball configuration and testing

### PR DESCRIPTION
**Pull request by Ibrahimmaga**
Beachball testing Scenarios
# Description
The testing will provide end to end testing with beachball

1. test for non change file
2. test for files that can trigger beachball
3. check for changes in package.json
4. ci-workflow

Adds and updates beachball package and configuration to project to handle the publishing of NPM public node packages on a nightly cadence through the CLI.

Files reside in `./build/releasing/*`, `package.json`, and `.github/workflows/ci-weekly.yml`

# Testing locally
Using scripts as noted in `package.json`  (@awentzel to test)
[x] 1. change any files and stage the commit  (@awentzel to test)
[x] 2. run the commands for `yarn change`, `yarn check`.  (@awentzel to test)
[x] 3. run the commands for `yarn publish` and `yarn public-ci` (@awentzel to test)
[] 3. run the commands for `yarn publish` and `yarn public-ci` (@chrisdholt   to test)

# Testing CI
[x] 1. test connection to GitHub API using PAT  (@awentzel to test)
[x] 2. test run tasks passes until NPM publish task  (@awentzel to test)
[ ] 3. change `ci-nightly.yml` node task to match that in package.json `yarn publish-ci` (@chrisdholt  to test)
[ ] 3. change `ci-nightly.yml` to use `on` `push` to master and on master branch (@chrisdholt  to test)

#Final CI test and PR change prior to merging
[ ] 1. change `ci-nightly.yml` to use `on` cron for master branch then commit this PR  (@chrisdholt  to test)

## Issue type checklist

- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [ ] **New feature**: A change that adds functionality.

**Is this a breaking change?**
- [ ] This change causes current functionality to break.

**Adding or modifying component(s) in `@microsoft/fast-components` checklist**

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## Process & policy checklist

- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### CODE REVIEW

**Comment by nicholasrice** [comment]
@Ibrahimmaga is this PR ready for review? If not, could you convert it to a draft PR so folks know to ignore it for now?

**Comment by Ibrahimmaga** [comment]
> @Ibrahimmaga is this PR ready for review? If not, could you convert it to a draft PR so folks know to ignore it for now?

yes you can review

File: change/@microsoft-fast-components-9902ce21-7d35-46de-8c68-28f33a5e69de.json
Comment (chrisdholt) [inline change/@microsoft-fast-components-9902ce21-7d35-46de-8c68-28f33a5e69de.json:3]: nit: title* (spelling).

File: change/@microsoft-fast-components-9902ce21-7d35-46de-8c68-28f33a5e69de.json
Comment (chrisdholt) [inline change/@microsoft-fast-components-9902ce21-7d35-46de-8c68-28f33a5e69de.json:2]: Is this an example or a test? If not, I think we'll want this to be "none"

Comment (Ibrahimmaga) [reply change/@microsoft-fast-components-9902ce21-7d35-46de-8c68-28f33a5e69de.json:2]: Just for testing purpose

**Comment by nicholasrice** [comment]
@Ibrahimmaga okay great, thanks! Can you update the PR title to align to our PR naming conventions?

File: .github/workflows/ci-daily.yml
Comment (chrisdholt) [inline .github/workflows/ci-daily.yml:40]:
```suggestion
    - name: Check for change files
```
A bit of a nit, but I think we should be specific here. This checks for the presence of "change" files for the purpose of ensuring we have metadata necessary to generate a changelog and version correctly. So, not technically "file changes" as the PR has that *always*, we really want to look for the specific files.

File: package.json
Comment (nicholasrice) [inline package.json:45]: Does beachball actually make commits to master like Lerna does?

Comment (Ibrahimmaga) [reply package.json:45]: I have not used Learna before. After you fill out the scope for file changed, it will automatically commit. But always check with git status to make sure.

File: package.json
Comment (chrisdholt) [inline package.json:42]: I'm checking, but I think we'll want to remove the `!` before "packages". I *think* that is saying, "not that folder".

File: package.json
Comment (chrisdholt) [reply package.json:45]: @nicholasrice this link presents the workflow in detail: https://microsoft.github.io/beachball/overview/getting-started.html

File: package.json
Comment (chrisdholt) [reply package.json:42]: Yeah, I think we want to have `\"packages/*\"`: https://github.com/microsoft/beachball/blob/master/src/monorepo/getScopedPackages.ts#L12

Comment (Ibrahimmaga) [reply package.json:42]: makes sense

File: change/@microsoft-fast-components-9902ce21-7d35-46de-8c68-28f33a5e69de.json
Comment (chrisdholt) [reply change/@microsoft-fast-components-9902ce21-7d35-46de-8c68-28f33a5e69de.json:2]: Cool - I'm going to "hold" while we test just to prevent this from being merged *with this in*. I think once we can confirm this is working as expected we'll need to publish w/ Lerna, merge this immediately after, and that will start enforcing these files. 👍

**Review by chrisdholt — CHANGES_REQUESTED** [review]
Holding on this while we test to ensure that the "test" change files don't accidentally get checked in :)

It won't get checked. bug good to block just in case.

**Comment by Ibrahimmaga** [comment]
> Are we at a point where we should set up a meeting with the team to go over procedures? Also, I think there was an old PR with documentation updates related to this...

Yes. I will go over this with Aaron in a few minutes, then we will update you

File: package.json
Comment (nicholasrice) [reply package.json:45]: @chrisdholt So yes? What I'm trying to understand is if there is a race condition between someone manually merging a PR prior and beachball's publishing commit.

Comment (chrisdholt) [reply package.json:45]: I think there could be, though it seems like that's acknowledged and they try to account for that discrepancy (sorry for another link): https://microsoft.github.io/beachball/cli/publish.html

> In most version bumping systems that automates git repo version with npm registry, they assume that the source code is fresh by the time they push the changes back to the git repository. This is almost never the case when the build takes more than a few minutes! So, beachball fetches latest before we push back to the target branch. In large systems, it has been observed that without the git fetch, it becomes a source of conflict.

FWIW (which may be very little), we've been using this on the Fluent UI side with both automated nightly publishes as well as manual publishes during the day. We haven't hit an issue with those *yet*...but I acknowledge, *yet*.

Comment (nicholasrice) [reply package.json:45]: Gotcha okay, well hopefully in practice it won't be an issue.

Comment (chrisdholt) [reply package.json:45]: @nicholasrice totally agree :). I do think there are a couple things which limit the risk (though there isn't inherently *no risk*):

1. Only contributors with write access have merge permissions. We can manage through policy in terms of *manual* publishes...if we we need to manually publish we ask folks to hold PR's.

2. We setup the nightly publish at a time which is *inconvenient* for merging anyways. Perhaps something like 2AM PST (5AM ET), etc...

Comment (chrisdholt) [reply package.json:45]: For 2, I think folks who *can merge* would be aware of this and ideally try and avoid merging at that time

Comment (chrisdholt) [reply package.json:45]: Lastly (I hope), If we see a manual publish has already been kicked off - we don't run it again while it's already going. 

I think those simple guidelines would help avoid the issue

File: .github/workflows/ci-daily.yml
Comment (awentzel) [reply .github/workflows/ci-daily.yml:40]: We could also include "Check for the presence of changed files inside `./change`".  The entire detail should be added to documentation we wrote today @Ibrahimmaga.

File: .github/workflows/ci-nightly.yml
Comment (awentzel) [inline .github/workflows/ci-nightly.yml:7]: I would still prefer this run once nightly unless this would pose an issue.  How do you feel about this, @chrisdholt?   If we're all in agreement, then we should replace the `pull_request` event with a cron schedule and only for the master branch.

I do think that in our `ci-daily.yml` we can include the `yarn checkchange` to provide indication for contributors that more work is required before pushing to master to prepare for publishing.

**Review by awentzel — COMMENTED** [review]
Please add the documentation, plus any additional details as Chris has noted, to the https://www.fast.design/docs/community/contributor-guide for awareness.

**Comment by awentzel** [comment]
@Ibrahimmaga please update this PR title to describe what change is being implemented instead of your branch name.  This is useful for when a changelog file is created, the community can scan and quickly understand how FAST is improving.

File: .github/workflows/ci-nightly.yml
Comment (awentzel) [inline .github/workflows/ci-nightly.yml:18]: PACK_TOKEN is not correct, use the one as shown for secret npm.

File: .github/workflows/ci-nightly.yml
Comment (chrisdholt) [reply .github/workflows/ci-nightly.yml:7]: I think I'm tracking - I would expect publish to *only* run on a cron with a method of manually triggering the build in emergencies or urgent cases only. *Ideally* a manual deployment would be triggered via an ADO build agent behind permissions or something similar so that it doesn't require any individual to run it from their own CLI.

+ @EisenbergEffect if he has thoughts as well

**Review by awentzel — APPROVED** [review]

File: .github/workflows/ci-nightly.yml
Comment (awentzel) [reply .github/workflows/ci-nightly.yml:7]: @chrisdholt, confirmed, we will be converting this to use Cron as soon as it's approved and as a final step before completing the PR.  We can put this in now if you want?

Comment (chrisdholt) [reply .github/workflows/ci-nightly.yml:7]: @awentzel let's include as part of this PR if we can. 

Once this goes in, every current/existing PR which includes files under `/packages/*` *should* fail at `yarn checkchange`. Once this is ready, we should pause merging and publish. Immediately after the publish is complete, we should merge this in. That kicks off the official shift from publishing w/ Lerna to publishiing w/ Beachball.

Comment (EisenbergEffect) [reply .github/workflows/ci-nightly.yml]: My original thought was to use NPM release channels.

* Canary Channel: Either triggered on every merge to master or by a nightly cron job. I don't have a strong opinion on which one. Should be at least once per day though.
* Beta Channel: Released once per week via a cron job.
* Stable Channel: Manually released by Chris, myself or other DRI at a planned time, typically in conjunction with messaging (e.g. blog post, tweets, conference, etc.).

Here's some information: https://github.com/danillouz/til-npm-release-channels
I think the biggest question was how to make this work with beachball.

File: packages/web-components/fast-components/src/listbox/scenarios/index.html
Comment (nicholasrice) [inline packages/web-components/fast-components/src/listbox/scenarios/index.html:5]: is this change related?

Comment (nicholasrice) [reply packages/web-components/fast-components/src/listbox/scenarios/index.html:5]: Ah nm I see the change test

Comment (Ibrahimmaga) [reply packages/web-components/fast-components/src/listbox/scenarios/index.html:5]: yes. we can leave the change if that okay. I added option 4 or just remove it. I believe no harm was done on this.

File: .github/workflows/ci-nightly.yml
Comment (Ibrahimmaga) [reply .github/workflows/ci-nightly.yml]: right now it is set to trigger 1 publish per night on cron job. Chris will know a little more.

File: package.json
Comment (chrisdholt) [inline package.json:42]: I was playing around with this locally - I think what we want is below. It should check for the presence of change files in all packages but not the sites, which would be what we want IMO. If there are changes *only* to sites, we don't need change files. Any changes under packages AND those would require them.

```suggestion
    "checkchange": "beachball check --scope \"!sites/*\" --changehint \"Run 'yarn change' to generate a change file\"",
```

Comment (chrisdholt) [reply package.json:42]: For some reason passing in packages/ explicitly doesn't work, so I the above is correct.

File: .github/workflows/ci-nightly.yml
Comment (chrisdholt) [reply .github/workflows/ci-nightly.yml]: I think we'll want to ensure this meets Rob's expectations as well. We definitely can use a nightly publish. If we need to release those on the NPM cadence, I'm not sure this work accommodates that.

Comment (awentzel) [reply .github/workflows/ci-nightly.yml]: I'd prefer to first get our publishing working and testing as an automated replacement of today's manual process.  Once that's in place and verified for a week or so we can most certainly discuss NPM channels.

- We may require contributing to beachball to get semver 2 meta data feature which to my knowledge has not been completed, such that we can properly tag the channel per previous GH Issues
- We may want to wait for GH/NPM Release capability to update.

In either case, that would be a good next step to discuss after we get the base automation in place.

Comment (EisenbergEffect) [reply .github/workflows/ci-nightly.yml]: Sounds good. Just want to make sure the goal is clear.

**Pull request merged**
